ddtrace/internal/_file_queue.py

```diff
@@ -0,0 +1,63 @@
+import os
+import os.path
+import tempfile
+import typing
+
+from ddtrace.internal._unpatched import unpatched_open
+
+
+try:
+    # Unix based file locking
+    # Availability: Unix, not Emscripten, not WASI.
+    import fcntl
+
+    def lock(f):
+        fcntl.lockf(f, fcntl.LOCK_EX)
+
+    def unlock(f):
+        fcntl.lockf(f, fcntl.LOCK_UN)
+
+except ModuleNotFoundError:
+    # Availability: Windows
+    import msvcrt
+
+    def size(f):
+        return os.path.getsize(os.path.realpath(f.name))
+
+    def lock(f):
+        msvcrt.locking(f.fileno(), msvcrt.LK_RLCK, size(f))
+
+    def unlock(f):
+        msvcrt.locking(f.fileno(), msvcrt.LK_UNLCK, size(f))
+
+
+class File_Queue:
+    """A simple file-based queue implementation for multiprocess communication."""
+
+    def __init__(self) -> None:
+        self.directory = tempfile.gettempdir()
+        self.filename = os.path.join(self.directory, "file_queue")
+
+    def put(self, data: str) -> None:
+        """Push a string to the queue."""
+        try:
+            with unpatched_open(self.filename, "a") as f:
+                lock(f)
+                f.write(data + "\x00")
+                unlock(f)
+        except Exception:  # nosec
+            pass
+
+    def get_all(self) -> typing.Set[str]:
+        """Pop all unique strings from the queue."""
+        try:
+            with unpatched_open(self.filename, "r+") as f:
+                lock(f)
+                data = f.read()
+                os.unlink(self.filename)
+                unlock(f)
```

Comment:
if we unlink before locking, we don't need to lock at all when reading. 
There is no perfect solution here I think, but the reader is single and won't read often, while the writers may be many and will write often. It's acceptable to have a few data losses for this feature, as it will be sent regularly to the backend.

Reply:
While looking for a solution for windows, I improved the whole thing. It should be perfect now. We don't delete the file, just truncate it to 0 length. after locking, we ensuring to seek at the right position for the operation.